### PR TITLE
*: fix compatibility with latest go-eth2-client

### PR DESCRIPTION
**Pull request by xenowits**
Fix charon's compatibility with latest release of `go-eth2-client` which contains breaking changes, notably the ETH2 client interface methods. This PR also enables easy integration of deneb fork as the data structures for deneb are included in the new release. An example of the API change is as follows:

Earlier
```
validatorsProvider.Validators(ctx, "head", nil)
```
Now
```
validatorsProvider.Validators(ctx, &api.ValidatorsOpts{
  State: "head",
})
```

Also fix corresponding tests.

category: feature 
ticket: #2646

### CODE REVIEW

**Comment by xenowits** [comment]
FYI: This PR also solves https://github.com/ObolNetwork/charon/issues/2477

File: core/fetcher/fetcher.go [hunk 1]
```diff
@@ -127,11 +127,21 @@ func (f *Fetcher) fetchAttesterData(ctx context.Context, slot int64, defSet core
 
 		eth2AttData, ok := dataByCommIdx[commIdx]
 		if !ok {
-			var err error
-			eth2AttData, err = f.eth2Cl.AttestationData(ctx, eth2p0.Slot(uint64(slot)), commIdx)
+			var (
+				err      error
+				eth2Resp *eth2api.Response[*eth2p0.AttestationData]
```
Comment (dB2510) [inline core/fetcher/fetcher.go:132]: any reason behind defining `eth2Resp` here before calling the endpoint?
we can do what we are already doing on line 218 in the same file.

File: core/signeddata.go [hunk 2]
```diff
@@ -201,148 +200,176 @@ func NewPartialVersionedSignedBeaconBlock(block *eth2spec.VersionedSignedBeaconB
 	}, nil
 }
 
-// VersionedSignedBeaconBlock is a signed versioned beacon block and implements SignedData.
-type VersionedSignedBeaconBlock struct {
-	eth2spec.VersionedSignedBeaconBlock // Could subtype instead of embed, but aligning with Attestation that cannot subtype.
+type VersionedSignedProposal struct {
```
Comment (dB2510) [inline core/signeddata.go:203]: godoc please

File: core/signeddata.go [hunk 3]
```diff
@@ -433,73 +460,33 @@ func NewPartialVersionedSignedBlindedBeaconBlock(block *eth2api.VersionedSignedB
 	}, nil
 }
 
-// VersionedSignedBlindedBeaconBlock is a signed versioned blinded beacon block and implements SignedData.
-type VersionedSignedBlindedBeaconBlock struct {
-	eth2api.VersionedSignedBlindedBeaconBlock // Could subtype instead of embed, but aligning with Attestation that cannot subtype.
+type VersionedSignedBlindedProposal struct {
```
Comment (dB2510) [inline core/signeddata.go:463]: godoc please

File: testutil/validatormock/propose.go [hunk 4]
```diff
@@ -138,11 +149,11 @@ func ProposeBlock(ctx context.Context, eth2Cl eth2wrap.Client, signFunc SignFunc
 			Message:   block.Capella,
 			Signature: sig,
 		}
-	default:
+	default: // TODO(xenowits): Add a case for deneb block.
```
Comment (dB2510) [inline testutil/validatormock/propose.go:152]: why this is not added in this PR as you have already added case for deneb almost everywhere?

File: testutil/validatormock/propose.go [hunk 5]
```diff
@@ -206,10 +222,16 @@ func ProposeBlindedBlock(ctx context.Context, eth2Cl eth2wrap.Client, signFunc S
 	}
 
 	// Get Unsigned beacon block with given randao and slot
-	block, err = eth2Cl.BlindedBeaconBlockProposal(ctx, slot, randao, nil)
+	proposalOpts := &eth2api.BlindedProposalOpts{
+		Slot:         slot,
+		RandaoReveal: randao,
+		Graffiti:     [32]byte{},
```
Comment (dB2510) [inline testutil/validatormock/propose.go:228]: nit: can omit this field

File: testutil/validatormock/propose.go [hunk 6]
```diff
@@ -89,10 +94,16 @@ func ProposeBlock(ctx context.Context, eth2Cl eth2wrap.Client, signFunc SignFunc
 	}
 
 	// Get Unsigned beacon block with given randao and slot
-	block, err = eth2Cl.BeaconBlockProposal(ctx, slot, randao, nil)
+	proposalOpts := &eth2api.ProposalOpts{
+		Slot:         slot,
+		RandaoReveal: randao,
+		Graffiti:     [32]byte{},
```
Comment (dB2510) [inline testutil/validatormock/propose.go:100]: nit: can omit this field

File: testutil/validatormock/propose.go [hunk 7]
```diff
@@ -247,11 +269,11 @@ func ProposeBlindedBlock(ctx context.Context, eth2Cl eth2wrap.Client, signFunc S
 			Message:   block.Capella,
 			Signature: sig,
 		}
-	default:
+	default: // TODO(xenowits): Add a case for deneb block.
```
Comment (dB2510) [inline testutil/validatormock/propose.go:272]: same here

File: core/validatorapi/validatorapi.go [hunk 8]
```diff
@@ -446,17 +407,58 @@ func (c Component) BlindedBeaconBlockProposal(ctx context.Context, slot eth2p0.S
...
 	if err != nil {
 		return nil, err
 	}
 
-	return block, nil
+	return &eth2api.Response[*eth2api.VersionedBlindedProposal]{Data: proposal}, nil
+}
+
+func (c *Component) SubmitProposal(ctx context.Context, proposal *eth2api.VersionedSignedProposal) error {
+	// Calculate slot epoch
```
Comment (dB2510) [inline core/validatorapi/validatorapi.go:419]: where epoch is calculated? seems like an old comment, can remove

File: core/validatorapi/validatorapi.go [hunk 9]
```diff
@@ -446,17 +407,58 @@ func (c Component) BlindedBeaconBlockProposal(ctx context.Context, slot eth2p0.S
...
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
 }
 
-func (c Component) SubmitBlindedBeaconBlock(ctx context.Context, block *eth2api.VersionedSignedBlindedBeaconBlock) error {
+func (c *Component) SubmitBlindedProposal(ctx context.Context, proposal *eth2api.VersionedSignedBlindedProposal) error {
```
Comment (dB2510) [inline core/validatorapi/validatorapi.go:459]:
```suggestion
func (c Component) SubmitBlindedProposal(ctx context.Context, proposal *eth2api.VersionedSignedBlindedProposal) error {
```

File: core/validatorapi/validatorapi.go [hunk 10]
```diff
@@ -446,17 +407,58 @@ func (c Component) BlindedBeaconBlockProposal(ctx context.Context, slot eth2p0.S
...
+	proposal, err := c.awaitBlindedProposalFunc(ctx, opts)
 	if err != nil {
 		return nil, err
 	}
 
-	return block, nil
+	return &eth2api.Response[*eth2api.VersionedBlindedProposal]{Data: proposal}, nil
+}
+
+func (c *Component) SubmitProposal(ctx context.Context, proposal *eth2api.VersionedSignedProposal) error {
```
Comment (dB2510) [inline core/validatorapi/validatorapi.go:418]:
```suggestion
func (c Component) SubmitProposal(ctx context.Context, proposal *eth2api.VersionedSignedProposal) error {
```

File: app/eth2wrap/synthproposer.go [hunk 11]
```diff
@@ -83,93 +88,115 @@ func (h *synthWrapper) SubmitProposalPreparations(ctx context.Context, preparati
...
 	}
 
-	block, err := h.syntheticBlock(ctx, slot, vIdx)
+	synthBlindedProposal, err := blindedProposal(proposal)
 	if err != nil {
 		return nil, err
 	}
 
-	return blindedBlock(block)
+	return &eth2api.Response[*eth2api.VersionedBlindedProposal]{Data: synthBlindedProposal}, nil
```
[the remark this answers is not in the document]
Comment (xenowits) [reply app/eth2wrap/synthproposer.go:137]: cool, added a function to wrap data into api response to wherever possible: beaconmock, validatorapi, synthproposer

File: core/fetcher/fetcher.go (hunk 1 above)
Comment (xenowits) [reply core/fetcher/fetcher.go:132]: yeah, modified

File: core/unsigneddata.go [hunk 12]
```diff
@@ -265,57 +264,57 @@ func (b *VersionedBeaconBlock) UnmarshalJSON(input []byte) error {
...
 	default:
-		return VersionedBlindedBeaconBlock{}, errors.New("unknown version")
+		return VersionedBlindedProposal{}, errors.New("unknown version")
 	}
 
-	return VersionedBlindedBeaconBlock{VersionedBlindedBeaconBlock: *block}, nil
+	return VersionedBlindedProposal{VersionedBlindedProposal: *proposal}, nil
+}
+
+type VersionedBlindedProposal struct {
```
Comment (gsora) [inline core/unsigneddata.go:294]: Missing godoc here

File: testutil/fuzz.go [hunk 13]
```diff
@@ -104,38 +104,58 @@ func NewEth2Fuzzer(t *testing.T, seed int64) *fuzz.Fuzzer {
...
+			// Populate one of the versions of these Versioned*Proposal types.
+			func(e *core.VersionedSignedBlindedProposal, c fuzz.Continue) {
 				e.Version = blindedVersions[(c.Intn(len(blindedVersions)))]
 				version, err := eth2util.DataVersionFromETH2(e.Version)
 				require.NoError(t, err)
 
 				val := core.VersionedSSZValueForT(t, e, version)
 				c.Fuzz(val)
+
+				maxBlobSidecars := 6 // Limit length of blob sidecars to 6
```
Comment (gsora) [inline testutil/fuzz.go:116]: Is this value defined in the spec?

File: testutil/fuzz.go [hunk 14]
```diff
@@ -104,38 +104,58 @@ func NewEth2Fuzzer(t *testing.T, seed int64) *fuzz.Fuzzer {
...
+				if e.Version == eth2spec.DataVersionDeneb && len(e.Deneb.SignedBlindedBlobSidecars) > maxBlobSidecars {
```
Comment (gsora) [inline testutil/fuzz.go:117]: Could we abstract this in a function?

File: testutil/validatormock/propose.go (hunk 4 above)
Comment (xenowits) [reply testutil/validatormock/propose.go:152]: yeah intended to, and I then got lazy. I have added it now for both `ProposeBlock` & `ProposeBlindedBlock`

File: core/validatorapi/router.go [hunk 15]
```diff
@@ -599,10 +649,16 @@ func proposeBlindedBlock(p eth2client.BlindedBeaconBlockProposalProvider) handle
 			return nil, nil, err
 		}
 
-		block, err := p.BlindedBeaconBlockProposal(ctx, eth2p0.Slot(slot), randao, nil)
+		opts := &eth2api.BlindedProposalOpts{
+			Slot:         eth2p0.Slot(slot),
+			RandaoReveal: randao,
+			Graffiti:     [32]byte{},
```
Comment (gsora) [inline core/validatorapi/router.go:655]: This can be omitted

File: testutil/fuzz.go (hunk 13 above)
Comment (xenowits) [reply testutil/fuzz.go:116]: yes, I'll add the link as a comment: https://github.com/ethereum/consensus-specs/blob/dev/specs/deneb/beacon-chain.md#execution

File: testutil/fuzz.go (hunk 14 above)
Comment (xenowits) [reply testutil/fuzz.go:117]: yes can be done, but I don't think it's worth it introducing a function for it. Moreover, there are 4 different types to deal with: Deneb.SignedBlindedBlobSidecars, Deneb.BlindedBlobSidecars, Deneb.SignedBlobSidecars, Deneb.BlobSidecars

File: core/dutydb/memory.go [hunk 16]
```diff
@@ -155,15 +154,15 @@ func (db *MemDB) Store(_ context.Context, duty core.Duty, unsignedSet core.Unsig
 	return nil
 }
 
-// AwaitBeaconBlock implements core.DutyDB, see its godoc.
-func (db *MemDB) AwaitBeaconBlock(ctx context.Context, slot int64) (*eth2spec.VersionedBeaconBlock, error) {
+// AwaitProposal implements core.DutyDB, see its godoc.
+func (db *MemDB) AwaitProposal(ctx context.Context, opts *eth2api.ProposalOpts) (*eth2api.VersionedProposal, error) {
```
[the remark this answers is not in the document]
Comment (xenowits) [reply core/dutydb/memory.go]: refactored to use only slot

File: core/dutydb/memory.go [hunk 17]
```diff
@@ -180,15 +179,15 @@ func (db *MemDB) AwaitBeaconBlock(ctx context.Context, slot int64) (*eth2spec.Ve
 	}
 }
 
-// AwaitBlindedBeaconBlock implements core.DutyDB, see its godoc.
-func (db *MemDB) AwaitBlindedBeaconBlock(ctx context.Context, slot int64) (*eth2api.VersionedBlindedBeaconBlock, error) {
+// AwaitBlindedProposal implements core.DutyDB, see its godoc.
+func (db *MemDB) AwaitBlindedProposal(ctx context.Context, opts *eth2api.BlindedProposalOpts) (*eth2api.VersionedBlindedProposal, error) {
```
[the remark this answers is not in the document]
Comment (xenowits) [reply core/dutydb/memory.go]: refactored to use only slot

File: app/eth2wrap/synthproposer_test.go [hunk 18]
```diff
@@ -59,11 +60,13 @@ func TestSynthProposer(t *testing.T) {
 	}
 	signedBeaconBlock := bmock.SignedBeaconBlock
 	bmock.SignedBeaconBlockFunc = func(ctx context.Context, blockID string) (*spec.VersionedSignedBeaconBlock, error) {
-		if rand.Float32() < 0.3 { // Fail to find 2/3 of blocks.
```
[the remark this answers is not in the document]
Comment (xenowits) [reply app/eth2wrap/synthproposer_test.go:62]: Not needed anymore, go-eth2-client fixed the nil,nil return response

**Review by dB2510 — APPROVED** [review]

**Review by gsora — APPROVED** [review]

**Pull request merged**